### PR TITLE
Initial ray tracing/query implementation

### PR DESCRIPTION
**Pull request by vudu2007**
### Ray Tracing
The Khronos **ray tracing extension** has been implemented. The features do not work as specifications until there is a way to run multiple different shaders in a single invocation of the interpreter.
- The ray generation execution model currently returns a payload that indicates whether the ray intersected a primitive; it does not execute a ray tracing pipeline yet.
- The examples only check that the instructions work and are not examples that would be used in producing a proper ray traced scene.
- The format is currently in a `*.json` format. The current acceleration structure (AS) input format contains all acceleration structures that will be used. At the very top is the metadata with information on what each acceleration structure contains. The top-level acceleration structure (TLAS) is at the beginning of all AS and the bottom-level acceleration structures (BLAS) follow.
- The interpreter will take in the AS format and convert it into an object-based form where the acceleration structure is constructed bottom-up.
- Currently, the acceleration structure is represented as a `AccelerationStructureManager` object that consist of `AccelerationStructure` objects that consists of `Node` objects.
- The `AccelerationStructureManager` contains the root `AccelerationStructure` (the TLAS) and the acceleration structure input formatted with the interpreter's data types, specifically `Aggregate`.
- Some of the added examples will contain scene.txt files that have Desmos links that provide a visual.

### Ray Query
The Khronos **ray query extension** is in progress.

### CODE REVIEW

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:33]: You don't need to specify a blank string. Simply use:

std::stringstream result;

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:32]: Should probably pass str by reference instead of copying it.

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:39]: Should probably pass the vec by reference instead of copying it.

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:63]: Please use the default keyword here:

virtual ~Node() = default;

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:202]: If you are passing a bool which won't be modified, I don't see the need to pass it by reference. You should often pass primitive types by value instead.

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:233]: Comment is unnecessary. "ID" typically stands for identifier / identification.

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:503]: This method does not follow the established style. Please edit the formatting to match what you see in instruction.cxx (for example, look at Instruction::readOp). Thanks.

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:506]: There is a significant amount of trailing spaces here (and elsewhere in the commit). You can do:

sed -i 's/[ \t]*$//' FILE_NAME

where you replace FILE_NAME with the name (if in the same directory), or you should be able to use the file path.

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:536]: I am not generally a fan of including parameters as editable pseudo-returns. Please return a tuple or struct containing both !nodesToEval.empty() and did_intersect_geometry, as needed.

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:562]: What remains to be done here? There is a TODO but no explanation and significant amount of code beneath it.

File: src/values/ray-query.cxx
Comment (mmoult) [inline src/values/ray-query.cxx:37]: Please fill in details here.

File: examples/raytracing-rahit/shader.rahit
Comment (mmoult) [inline examples/raytracing-rahit/shader.rahit:9]: I don't understand this comment. What does the shader do differently to cooperate with the interpreter?

File: examples/raytracing-rgen/out9.json
Comment (mmoult) [inline examples/raytracing-rgen/out9.json:1]: Comment here since GitHub won't let me comment directly on the rgen.spv binary:

This example appears incomplete. I tried running:
spirv-run -t out.yaml rgen.spv

and I got an exception: "Cannot print value!". I got the same error when trying to run:
spirv-run -t out.json rgen.spv

You should support printing of template variables. This is a very important feature of SPIRV-Interpreter which differentiates it from other similar projects.

File: src/spv/inst-make.cpp
Comment (mmoult) [inline src/spv/inst-make.cpp:1218]: I think this comment can be improved. I remember you explaining this concept, but I don't remember all the details. Correct me if I am wrong, but from what I recall, the reason why we cannot define the type here is because the SPIR-V instruction doesn't give enough info (many fields of raytracing types are opaque), so we have to wait until a use. Right?

If that is the case, it may be wise to create a new pseudo constructor for acceleration structure types which doesn't take any arguments. It seems silly to me to pass in empty vectors here. Just let the type populate them itself.

File: src/spv/inst-make.cpp
Comment (mmoult) [inline src/spv/inst-make.cpp:1233]: Are these instructions used in your examples? If there is a place to put TODO, it would probably be here (to note that this code only generates dummy values). However, I would prefer you to get rid of half-implemented features.

File: src/spv/inst-read.cpp
Comment (mmoult) [inline src/spv/inst-read.cpp:402]: I agree with your TODOs. This seems to fit better by the enum class Extension in instruction.cxx. You could even make a static isExtensionPresent method or something, since I suspect we will use this behavior again.

**Review by mmoult — CHANGES_REQUESTED** [review]
Here are some initial thoughts I have. I will continue to look at the code to investigate the more complicated parts, but the issues I raise are something for you to get started with until I am done.

**Comment by mmoult** [comment]
Also notice that I created a new branch for in-progress raytracing commits. I didn't feel comfortable merging this to the main branch with all the debug prints and what not.

File: examples/raytracing-rgen/out17.json
Comment (vudu2007) [reply examples/raytracing-rgen/out17.json:2]: It was manually inputted.

File: src/values/acceleration-structure.cxx
Comment (vudu2007) [reply src/values/acceleration-structure.cxx:781]: Should add this clang format option, `AlignAfterOpenBracket: BlockIndent`, to `.clang-format`?

File: examples/raytracing-rint/scene.txt
Comment (vudu2007) [reply examples/raytracing-rint/scene.txt:1]: I've read that they are permalinks (https://help.desmos.com/hc/en-us/articles/4405901719309-Saving-and-Managing-Graphs). However, I will include scene data just in case.

File: examples/raytracing-rahit/shader.rahit
Comment (vudu2007) [reply examples/raytracing-rahit/shader.rahit:9]: GLSL clamp instruction wasn't implemented in the interpreter yet, so I placed that comment. Looking at it again, it looks confusing without context so I will remove it.

File: src/spv/inst-make.cpp
Comment (vudu2007) [reply src/spv/inst-make.cpp]: Correct! I've decided to use default arguments for creating an empty acceleration structure type.

File: src/spv/inst-read.cpp
Comment (vudu2007) [reply src/spv/inst-read.cpp:421]: Do you think extensions and extended instructions should be separate? Currently, the interpreter considers `GLSL.std.` an extension rather than an extended instruction, though I'm not sure about the differences yet.

File: src/spv/inst-read.cpp
Comment (vudu2007) [reply src/spv/inst-read.cpp:134]: It seems that this instruction if only for validation and there are no results from executing it. The only thing being done by the interpreter is checking if the extension is supported. Thus, it's probably only involved in `inst-read.cpp` so I'll just remove the comment.

File: src/values/acceleration-structure.cxx
Comment (mmoult) [reply src/values/acceleration-structure.cxx]: Give it a try. I don't have a ton of experience with the clang-format, so I am still tweaking it to match the code style I expect.

File: src/spv/inst-read.cpp
Comment (mmoult) [reply src/spv/inst-read.cpp:421]: The main difference between extensions and extended instructions is whether they are used like Capabilities (for the former) or a generic instruction extension. That is a fairly superficial distinction.

I have thought about this a lot and now I am on the fence whether we want to check extensions at all. We don't check Capabilities, for example, so it makes sense we wouldn't check extensions if they are not directly useful.

I made a design choice earlier not to worry too much about decorations and capabilities. There are many, they aren't always relevant, and SPIR-V continues to change, so it makes the interpreter more robust if it can ignore decorations it doesn't need. On the other hand, that opens the door for silent failures.

File: examples/raytracing-rint/shader.rint
Comment (vudu2007) [reply examples/raytracing-rint/shader.rint]: Replaced it with code that isn't directly copied and modified from Scratchapixel. The shader is still a sphere intersection test.

File: examples/raytracing-rint/scene.txt
Comment (vudu2007) [reply examples/raytracing-rint/scene.txt]: Ok, I added the variables and equations needed to visualize the scene in Desmos 3-D. I believe there were only two `scene.txt` files.

File: src/spv/inst-read.cpp
Comment (vudu2007) [reply src/spv/inst-read.cpp]: Looks like you're handling it in recent commits

File: src/spv/inst-read.cpp
Comment (vudu2007) [reply src/spv/inst-read.cpp:134]: Looks like you're handling it in recent commits

File: src/values/acceleration-structure.cxx
Comment (vudu2007) [reply src/values/acceleration-structure.cxx]: Hopefully, I got all the files I modified

File: examples/raytracing-rint/shader.rint
Comment (vudu2007) [reply examples/raytracing-rint/shader.rint]: I'm also using external code in my ray tracing intersection code. The AABB intersection is from Scratchapixel, so I'll definitely need to change that. The triangle intersection code was inspired by Scratchapixel, but not directly copied and modified, where the Moller-Trumbore ray-triangle intersection algorithm is used.

File: src/spv/inst-read.cpp
Comment (mmoult) [reply src/spv/inst-read.cpp]: Please remove this logic. I see that examples/opencl-simple is failing with error "Unsupported extension: SPV_KHR_storage_buffer_storage_class".

File: src/values/value.hpp
Comment (mmoult) [inline src/values/value.hpp:32]: This is good, but I think it belongs much better in Type's sameBase rather than here. Please move the check there.

File: src/values/type.cpp
Comment (mmoult) [inline src/values/type.cpp:61]: Trailing whitespace on this line. Please remove it on the next commit.

File: src/values/ray-query.cxx
Comment (mmoult) [inline src/values/ray-query.cxx:268]: The formatting is a little wonky for this line and 291, but I suspect that using emplace_back will fix that.

Note: whenever you find yourself pushing an object you just constructed, you should probably be using emplace_back (https://en.cppreference.com/w/cpp/container/vector/emplace_back)

File: src/values/ray-query.cxx
Comment (mmoult) [inline src/values/ray-query.cxx:278]: The way you are doing things here (get* calls get*GLM) works fine, but I think there is a way you can do it to reduce redundancy:

You could probably make a method which converts from Values to glm matrices (or vice versa). Then call that function from your individual get* methods and remove the for loop after which constructs the result vector.

File: src/values/ray-query.cxx
Comment (mmoult) [inline src/values/ray-query.cxx:286]: I see that you casted the input of getIntersectionWorldToObjectGLM to a bool, then the parameter for getIntersectionWorldToObject is an unsigned int again. I like that you casted it to a bool- please update the argument of getIntersectionWorldToObject.

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:1384]: Assuming you get something similar after the update, please replace this part with:

t_min = std::min(tz_min, t_min);

Comment (mmoult) [reply src/values/acceleration-structure.cxx:1384]: Sorry, it looks like I misread it- your code is finding the max, not the min. If you know what is going on, it may be helpful to give a comment since t_min makes it sound like the min is needed.

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:36]: We could be more clever about this since we know ahead of time how long the result will be (and therefore, we should be able to reserve that exact space), but I suppose this is all debug-only, so it shouldn't really matter.

File: src/spv/inst-exec.cpp
Comment (mmoult) [inline src/spv/inst-exec.cpp:450]: This seems wrong. You should not be reaching into the primitive to set its raw value directly. One complication is that the result type may have a different precision than you are expecting, and setting it directly fails to account for that.

Please use Primitive's copyFrom. Granted, I don't have the precision handling code in yet, but that is where it will be.

Please do the same for all other instances. I think it is even safer to do copyFrom for the boolean case just in case the primitive isn't actually a boolean (say in the future Khronos says we can use unsigned ints as the result- copyFrom would handle that).

File: examples/rayquery/shader.frag
Comment (mmoult) [inline examples/rayquery/shader.frag:1]: Please remove this example from the PR (and include it in the next one). It still seems incomplete and is giving me an error when I try to run it (with the input):
Cannot copy from struct of a different size (3 -> 8)!

There are already a lot of examples in this PR, so it would be nice to remove some content which seems unfinished.

File: examples/raytracing-rahit/shader.rahit
Comment (mmoult) [inline examples/raytracing-rahit/shader.rahit:15]: This example is interesting because ignoreIntersectionEXT and terminateRayEXT don't do anything presently. Maybe we can keep this example for the time being because it provides another shader stage, but it doesn't make sense to keep it independent indefinitely if it needs to run a closest hit shader or what not.

File: examples/raytracing-rchit/shader.rchit
Comment (mmoult) [inline examples/raytracing-rchit/shader.rchit:6]: Are `hitAttributeEXT`s always read-only, or just this particular one? If they are always read-only, then they shouldn't be included as part of the out interface. If this specific instance is only read, then I think that is obvious from how small the shader is. In either case, I think you should remove the comment.

Comment (mmoult) [reply examples/raytracing-rchit/shader.rchit:6]: I found this, about `HitAttributeNV`:

> They can be written to only in IntersectionNV execution model and read from only in AnyHitNV and ClosestHitNV execution models.

I recommend adding another parameter to `ioGen` which saves the shader type. This should help us determine whether the variable is readable/writeable.

File: examples/raytracing-rgen/shader.rgen
Comment (mmoult) [inline examples/raytracing-rgen/shader.rgen:19]: Is this value typically a double? The interpreter doesn't actually support doubles yet (it silently treats them as floats), so this is something for me to put higher on the to do list if so.

Also, be aware of this if you add other tests to this example or aren't seeing the precision you are expecting.

I am happy to have a real use case for double values, so please don't change this!

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:1434]: Please replace with:
`bool intersect_front = determinant >= epsilon;`

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:639]: Unless I am mistaken, this method, `CandidateIntersection::operator=`, and `CommittedIntersection::operator=` match the default-generated `operator=` exactly. I would suggest omitting these redundant definitions.

**Review by mmoult — CHANGES_REQUESTED** [review]
Very good work. The PR is almost complete. Please address the new comments I left, and I think it will be ready to merge.

I am also going to move this back to main (instead of the temporary raytracing branch) since it feels more well-rounded now (and since you are pulling all the changes I make to the main branch). I am going to hold off on new commits to master (if possible) until this is merged so you don't have to worry about rebasing in the interim.

File: examples/raytracing-rgen/shader.rgen
Comment (vudu2007) [reply examples/raytracing-rgen/shader.rgen:19]: Will keep doubles in mind.

File: src/spv/inst-exec.cpp
Comment (mmoult) [inline src/spv/inst-exec.cpp:329]: nit: Please ad the opcode here and for 335.

File: src/values/acceleration-structure.cxx
Comment (mmoult) [inline src/values/acceleration-structure.cxx:19]: Definitely looks like a compile (not a link-time) dependency to me. Please remove this comment or explain what you meant.

File: src/values/ray-query.cxx
Comment (mmoult) [inline src/values/ray-query.cxx:124]: Please remove this unused function and getWorldRayDirectionGLM() unless you plan to use them in the next update.

File: src/values/ray-query.cxx
Comment (mmoult) [inline src/values/ray-query.cxx:199]: Personally, I would fold this and the other get*GLM methods into their get* counterparts. Or at least, do these need to be public? Also, they are currently only called by their get* counterpart, so I don't know if they justify their independence.

File: src/values/ray-query.cxx
Comment (mmoult) [reply src/values/ray-query.cxx:278]: Looks good!

**Review by mmoult — CHANGES_REQUESTED** [review]
As discussed.

File: src/values/acceleration-structure.cxx
Comment (vudu2007) [reply src/values/acceleration-structure.cxx:19]: I was thinking that reducing the dependencies would reduce compile time. I guess "link" was the wrong word? But if there are no issues with including unused GLM files, I'll just remove the comment.

File: src/values/ray-query.cxx
Comment (vudu2007) [reply src/values/ray-query.cxx:199]: Decided to merge getGLM into get methods.

**Pull request merged**
